ogc: restore blending mode after every frame

mardy opened the pull request:
The OGC_video_flip() function does not simply flip the current video frame: if the mouse cursor is visible, it will draw it. If/when we'll add support for the OSK keyboard, that will also be drawn at the end of every frame.

Since these operations require a specific blending mode, we must make sure that it is restored to its original value before starting drawing the next frame.

We could optimize this by having OGC_video_flip() return a boolean (or a bit mask, in case there are more settings beside the blending mode that it can alter) telling whether the blending mode was changed, in order not to perform an unneeded operation. It might make sense to do it once we decide to do the same for the vertex formats and the texture stages, which change more often and are heavier operations.

carstene1ns commented:
>It doesn't cause any issues because it's never used.

This is wrong, you just do not understand what it did. It was used for an easy way to make sdl-based games without joystick support work by mapping their keyboard buttons.

That being said, it would have been more sane to just remove the arbitrary mapping table and be done with it.

sergiou87 commented:
Oh, now I understand what you mean 😓 

I still think that's something that should be out of SDL. It was just a shortcut so games using this port didn't need to work on their own joystick button bindings, but that's beyond SDL's purpose, isn't it?

A different discussion is whether we're comfortable breaking compatibility with homebrew that used this already. In my opinion, they were using an API that shouldn't exist to begin with, so they should adapt.

Besides, I haven't seen such an API in ports to other platforms (maybe I missed it 😳) which means code that uses that API is not very portable.

This is just my opinion, and I'm open to roll that change back if you guys think it makes sense.

WinterMute commented:
Might be more productive to look at an SDL port that uses this functionality and see what we can do to fix it. Can you point out something that does that @carstene1ns 

Being honest, I think @sergiou87 is right and people shouldn't bodge up shortcuts that make games made with portable libraries less portable.

mardy committed:
ogc: restore blending mode after every frame

The OGC_video_flip() function does not simply flip the current video
frame: if the mouse cursor is visible, it will draw it. If/when we'll
add support for the OSK keyboard, that will also be drawn at the end of
every frame.

Since these operations require a specific blending mode, we must make
sure that it is restored to its original value before starting drawing
the next frame.

We could optimize this by having OGC_video_flip() return a boolean (or
a bit mask, in case there are more settings beside the blending mode
that it can alter) telling whether the blending mode was changed, in
order not to perform an unneeded operation. It might make sense to do it
once we decide to do the same for the vertex formats and the texture
stages, which change more often and are heavier operations.

WinterMute commented:
Nice commit messages btw. I do rather enjoy the detail here.
